Add Zephyr blog post on structuring Pulumi projects (#2548)
pulumi/pulumi-hugo: themes/default/content/blog/iac-recommended-practices-structuring-pulumi-projects/index.md

---
title: "Iac Recommended Practices: Structuring Pulumi Projects"

> date: 2023-03-16T16:50:38-07:00

Reviewer: ```suggestion
date: 2023-03-17
```

meta_desc: The third post in the Zephyr series examines recommended practices around structuring Pulumi projects.

> meta_image: meta.png

Author: Yep, fixed

authors:
    - scott-lowe
    - christian-nunciato
    - aaron-kao
tags:
    - best-practices
    - cloud-engineering
    - zephyr
---

> This is the third post in a series of blog posts focused on Zephyr Archaeotech Emporium---our fictional company---and their use of Pulumi to manage their online retail store. In the first post, you saw how Zephyr initially decided to go with a single Pulumi project for managing deployments of their online retail store application. In this post, you'll see how Zephyr's use of Pulumi evolves as their company grows and evolves.<!--more-->

Author: The duplication was intentional, to draw a parallel/connection that when the company evolves the recommendations for using Pulumi also evolve.

* **Company/department/team structure:** How a company/department/team is organized should certainly be taken into account when structuring your Pulumi projects. In Zephyr's early days with Pulumi, a single team managed everything so a single project was fine. Now with multiple teams, Zephyr will likely find using multiple projects to be more effective.
* **The relationship between resources:** The relationships between resources---such as which resources are dependent upon other resources---is another factor that feeds into your Pulumi project structure. A "shared infrastructure layer" is often best handled in a separate project of its own, but may pull in resources that might otherwise have been kept with an application in its own project. (In some respects, this is a "counter-balance" to the first bullet point.) On the other hand, resources that belong exclusively to a specific application or function might be best bundled in the same Pulumi project.

> * **Resource lifecycle:** Do resources have a shared lifecycle, meaning they are always created, updated, or deleted together? (Note this is _not_ the same as a dependency on another resource!) If so, then those resources should be placed in the same project. If the resources do have a shared lifecycle, then placing them in the same project simplifies lifecycle management. A great example of resources with a shared lifecycle would be a Kubernetes Deployment and the associated Kubernetes Service.

Author: I've just removed the sentence, since it seems to be distracting.

* An infrastructure project: This project handled the AWS VPC and related infrastructure resources.
* A platform project: This project managed the Kubernetes cluster(s).
* An application project: This project handled the actual online retail store itself.

{{% notes %}}
You can use a combination of code refactoring and `pulumi import` to assist in the process of restructuring Pulumi projects, but it some cases it may be easier to create new infrastructure and migrate over to it.
{{% /notes %}}

In [the first Zephyr post](/blog/iac-recommended-practices-code-organization-and-stacks/), you read about some of the options for where to store Pulumi code relative to the applications it supports. At that time, Zephyr had decided to go with what's called a monorepo approach, meaning that the Pulumi code was stored in the same repository as the application code. In conjunction with the project reorganization, Zephyr also decided to change how their code was organized:

* Code for the infrastructure project moved to a separate "zephyr-infra" repository

> * A new "zephyr-k8s" repository housed the code for the platform project

Reviewer: What would you think about calling these "network" and "platform"? You could imagine, for example, something other than Kubernetes acting as the deployment later but still residing in that same project, and the infra project looks like it's all about defining the networking. Thoughts?

Author: At this point renaming things will introduce additional delays that I'd rather avoid if possible. We can always show Zephyr using `pulumi stack rename` later and specifically call out to readers that this is how you rename projects and stacks to align with changes over time.

Note that relocating code (to a different GitHub repository or different filesystem location within a repository) generally has no impact on Pulumi. As long as the project name and stack name do not change, then changing the filesystem location won't affect anything---keeping in mind, of course, that filesystem paths and such may need to be adjusted in code to account for any such changes.

These repositories are available for you to review on GitHub; use the branch/tag selector to find the `multi-project` tab in each repository to see the state of the code as of this blog post.

## Summarizing recommended practices

This post covered the following recommended practices for working with Pulumi:

* **Use multiple Pulumi projects** when you need the added flexibility that multiple Pulumi projects offer.
* **Align your Pulumi projects** according to external factors like company/department/team structure, resource dependencies, lifecycle, and use case/application.
* **Change the filesystem location** for Pulumi code as needed without impacting the resources managed by that code, as long as the Pulumi project name and stack name don't change.

> The next post will focus on some specific technical details related to Zephyr's project reorganization; specifically, how they used StackReferences to share information between multiple Pulumi projects. Stay tuned!

Reviewer: Since we're talking specifically about these projects:

```suggestion
The next post will focus on some specific technical details related to Zephyr's project reorganization; specifically, how they used StackReferences to share information between these Pulumi projects. Stay tuned!
```

---
title: "Iac Recommended Practices: Structuring Pulumi Projects"
date: 2023-03-16T16:50:38-07:00
meta_desc: The third post in the Zephyr series examines recommended practices around structuring Pulumi projects.
meta_image: meta.png
authors:
    - scott-lowe
    - christian-nunciato
    - aaron-kao
tags:
    - best-practices
    - cloud-engineering
    - zephyr
---

This is the third post in a series of blog posts focused on Zephyr Archaeotech Emporium---our fictional company---and their use of Pulumi to manage their online retail store. In the first post, you saw how Zephyr initially decided to go with a single Pulumi project for managing deployments of their online retail store application. In this post, you'll see how Zephyr's use of Pulumi evolves as their company grows and evolves.<!--more-->

> The [first blog post](/blog/iac-recommended-practices-code-organization-and-stacks/) in the Zephyr series specifically mentions that the series will show "how Zephyr's use of Pulumi evolves as the company grows and their online retail store application changes to accommodate their growth." While the ultimate goal of this series is to discuss recommended practices for using Pulumi to manage a fairly complex containerized application, it's important to note that many recommended practices are "point in time" recommendations---and they will change over time as the companies using Pulumi also change. This blog post is an example of that evolution in action: what was right for Zephyr while they were smaller may not be right for Zephyr when they are larger.

Reviewer 2: ```suggestion
The [first blog post](/blog/iac-recommended-practices-code-organization-and-stacks/) in the Zephyr series specifically mentions that the series will show "how Zephyr's use of Pulumi evolves as the company grows and their online retail store application changes to accommodate their growth." While the ultimate goal of this series is to discuss recommended practices for using Pulumi to manage a fairly complex containerized application, it's important to note that many recommended practices are "point in time" recommendations: best practices are context-dependent and the recommended practices for a given organization will change as the company and its infrastructure needs also change. This blog post is an example of that evolution in action: the practices that worked well for Zephyr while they were smaller and had smaller infrastructure and less teams working on that infrastructure are no longer ideal as the company has grown.
```

Reviewer 2: General gist of the above suggestion is to avoid "they" "those" etc and explicitly restate the term for clarity. I think this is called getting rid of unqualified articles or something like that. I learned it in a college technical writing class and like the way it reads. YMMV.

Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published---this will get updated as new posts are published):

[IaC Recommended Practices: Code Organization and Stacks](/blog/iac-recommended-practices-code-organization-and-stacks/)

[IaC Recommended Practices: Developer Stacks and Git Branches](/blog/iac-recommended-practices-developer-stacks-git-branches/)

> **IaC Recommended Practices: Structuring Pulumi Projects** (this post)

Reviewer 2: Could even do "YOU ARE HERE" if you like the mood that evokes.

IaC Recommended Practices: Tying Stacks Together with Stack References

IaC Recommended Practices: Local Testing with Pulumi

IaC Recommended Practices: Evolving the Application

IaC Recommended Practices: Adding Pulumi Deployments

IaC Recommended Practices: Refactoring for Reuse

## Catching up with Zephyr

> As a result of launching its online presence, Zephyr saw amazing success. The market for arcane artifacts and novel curiosities is booming, and Zephyr is becoming a leader in this market. Funded by this success, Zephyr has grown rapidly. Along with this growth, Zephyr's software engineering and IT teams also grew and expanded. Fortunately, Zephyr's [use of per-developer stacks with short-lived ephemeral Git branches](/blog/iac-recommended-practices-developer-stacks-git-branches/) positioned the development teams well for this growth.

Reviewer 2: 👍 This reads real solid.

{{% notes %}}

> For a look at the code as it was at the end of the second post in the series, see the [`dev-stacks`](https://github.com/pulumi/zephyr-app/tree/dev-stacks) tag of the [`pulumi/zephyr-app`](https://github.com/pulumi/zephyr-app) repository on GitHub.

Reviewer 2: ```suggestion
For a look at the state of Zephyr's Pulumi code as it stood as of the end of the second post in the series, see the [`dev-stacks`](https://github.com/pulumi/zephyr-app/tree/dev-stacks) tag of the [`pulumi/zephyr-app`](https://github.com/pulumi/zephyr-app) repository on GitHub.
```

{{% /notes %}}

> Naturally this growth has resulted in some separation of duties. The IT team now has responsibility for the infrastructure that supports the online retail store. Within that team, some team members are focused on "core" infrastructure, while other team members are focused only on Kubernetes (a natural decision given Zephyr's use of Kubernetes in their architecture). The application team continues to remain responsible for the online store application itself.

Reviewer 2: ```suggestion
Naturally, Zephyr's growth has resulted in some separation of duties among the infrastructure teams: The IT team now has responsibility for the infrastructure that supports the online retail store. Within that team, some team members are focused on "core" infrastructure like networking and object storage, while other team members are focused only on Kubernetes (a natural decision given Zephyr's use of Kubernetes in their architecture). The application team continues to remain responsible for the online store application itself.
```

> This left Zephyr again asking some questions about the best path forward:

Reviewer 2: ```suggestion
This new division of responsibilities among the engineers has left Zephyr with some questions about the best path forward:
```

> * **Should they use a single project, or multiple projects?** Zephyr had used a single Pulumi project up until now, but growth and organizational change meant they wanted to make sure they were using Pulumi in a way that best reflected their needs. What are the advantages and disadvantages of each approach?

Reviewer 2: ```suggestion
* **Should Zephyr use a single Pulumi project, or multiple projects?** Zephyr had used a single Pulumi project up until now, but growth and organizational change meant they wanted to make sure they were using Pulumi in a way that best reflected their needs. What are the advantages and disadvantages of each approach?
```

* **If multiple projects, how many projects?** What's the best way to determine how many projects are needed? What factors play a role in this decision?

> ## Structuring Pulumi projects

Reviewer 2: I would include blast radius in this section.

Author: Resolved via e08be4c

> One project, or multiple projects? And if multiple projects, how many projects? These are the core questions the Zephyr needed to resolve. In the end, there are a number of key factors that determine the answers to these questions:

Reviewer 2: I might skip this paragraph since it almost directly duplicates the content above.

> * **Use case:** Zephyr has only a single application (so far!), but in situations where there are multiple applications then keeping those applications (and sometimes their associated infrastructure, depending on other factors) in separate projects is generally recommended.

Reviewer 2: ```suggestion
* **Use case:** Zephyr has only a single application (so far!), but in situations where teams or organizations are responsible for multiple applications, keeping those applications (and sometimes their associated infrastructure, depending on other factors) in separate projects is generally recommended.
```

Reviewer 2: I think this advice also needs to include our assumptions about what infra, if any, these multiple workloads share (networking immediately comes to mind).

> * **Company/department/team structure:** How a company/department/team is organized should certainly be taken into account when structuring your Pulumi projects. In Zephyr's early days with Pulumi, a single team managed everything so a single project was fine. Now with multiple teams, Zephyr will likely find using multiple projects to be more effective.

Reviewer 2: Need to include the repository boundary as well here. One great reason for separate repos is that repos provide a security boundary: the team that owns a repo have the admin rights, ability to approve PRs (or decide whether they want to accept them in the first place), etc.

> * **The relationship between resources:** The relationships between resources---such as which resources are dependent upon other resources---is another factor that feeds into your Pulumi project structure. A "shared infrastructure layer" is often best handled in a separate project of its own, but may pull in resources that might otherwise have been kept with an application in its own project. (In some respects, this is a "counter-balance" to the first bullet point.)

Reviewer 2: Don't need to repeat same text immediately after the bold text. Give an example here, e.g. an app service depends on a K8s cluster which in turn depends on a VPC.

> * **Resource lifecycle:** Do resources have a shared lifecycle, meaning they are always created, updated, or deleted together? (Note this is _not_ the same as a dependency on another resource!) If so, then those resources should be placed in the same project.

Reviewer 2: Again, example would be good here.

Author: @jkodroff An example for each point, an example for a specific point, or something else?

Author: Addressed in e08be4c, via adding a couple examples and rewording some of the sections

> * **Change rate of resources:** Some resources may change very infrequently. Consider the example of an AWS VPC and associated objects. Once it is created, it may not change for extended periods of time, if ever. On the other hand, some resources may change frequently (in other words, these resources have a high change rate). In a single project, combining resources that have a high change rate with resources that have a low change rate may be suboptimal.

Reviewer 2: If we mean "blast radius", we should use that term. (And blast radius is definitely a concern with repo structure.)

Author: We don't mean blast radius here; that's mentioned later in the article.

Now that you have an idea of some of the factors that affect how users structure their Pulumi projects, what are the advantages and disadvantages of each approach?

For single projects, the key advantage is _simplicity._ There's no denying that a single project is simpler and less complex than working with multiple projects. This advantage, though, is also the key disadvantage. A single project does not afford the flexibility to let different teams manage different aspects of the infrastructure, for example, nor does it offer the ability to split resources that don't have a shared lifecycle. You can't run `pulumi destroy` on only _some_ of the resources in a project.

For multiple projects, the key advantage is _flexibility._ It's true that the added flexibility comes at the cost of some added complexity (a future post will explore one aspect of that added complexity in the form of StackReferences). However, you gain the ability to separate infrastructure resources according to team ownership/responsibility, relationship to other resources (i.e., grouping resources in a "shared infrastructure layer"), or any of the other dimensions mentioned above.

> Finally, there is the consideration of what happens in the event of human error, and limiting the so-called "blast radius." With a single project, _all_ the resources in that project are within the blast radius---all of them could potentially be affected in the event of human error. Using separate Pulumi projects can limit the blast radius. It's not something that should drive the decision all by itself, but it _is_ a factor to be considered.

Reviewer 2: I think this should be raised earlier because it applies even when there's 1 person working in isolation - you still need to limit blast radius. In fact, it was the first principle I learned for structuring IaC projects many moons ago.

Author: @jkodroff Where would you place the discussion blast radius, if not here?

Reviewer 2: Left comment above.

Author: Addressed in e08be4c

In some ways, the decision points for single vs. multiple projects are analogous to the decision points for monolith vs. microservice: independent scaling and flexibility versus simplicity.

_In general, most customers end up using multiple Pulumi projects per use case._ As you've seen, though, there are a number of factors that influence that decision and will affect how many projects end up in use.

## Moving forward with Zephyr

Zephyr had a number of factors in play, not the least of which was the organizational change resulting from their growth and success. In the end, Zephyr decided to restructure their use of Pulumi for the online retail store into three separate projects:

* An infrastructure project: This project handled the AWS VPC and related infrastructure resources.
* A platform project: This project managed the Kubernetes cluster(s).
* An application project: This project handled the actual online retail store itself.

{{% notes %}}
Note that today, Pulumi has no supported method to move resources from one project to another, so project restructuring will involve creating new infrastructure and migrating over to it.

> {{% /notes %}}

Reviewer 3: What about importing an existing resource (with a `.get()`) and deleting it from the original stack?

Author: I tried a lot of different things, but in the end we decided the official statement we should make was this one. There _are_ ways to move resources, but they aren't necessarily supported ways.

Reviewer 2: `pulumi import` is supported, and I think it's a reasonable alternative to this form of "create before destroy". 

Keep in mind for stateful infra like DBs, it's not necessarily practical to say "spin up a new DB and restore a backup". The DB in question could be ginormous.

Author: Fixed in e08be4c, let me know what you think of the updated wording

---
title: "Iac Recommended Practices: Structuring Pulumi Projects"
date: 2023-03-16T16:50:38-07:00
meta_desc: The third post in the Zephyr series examines recommended practices around structuring Pulumi projects.
meta_image: meta.png
authors:
    - scott-lowe
    - christian-nunciato
    - aaron-kao
tags:
    - best-practices
    - cloud-engineering
    - zephyr
---

This is the third post in a series of blog posts focused on Zephyr Archaeotech Emporium---our fictional company---and their use of Pulumi to manage their online retail store. In the first post, you saw how Zephyr initially decided to go with a single Pulumi project for managing deployments of their online retail store application. In this post, you'll see how Zephyr's use of Pulumi evolves as their company grows and evolves.<!--more-->

> The [first blog post](/blog/iac-recommended-practices-code-organization-and-stacks/) in the Zephyr series specifically mentions that the series will show "how Zephyr's use of Pulumi evolves as the company grows and their online retail store application changes to accommodate their growth." While the ultimate goal of this series is to discuss recommended practices for using Pulumi to manage a fairly complex containerized application, it's important to note that these recommended practices can and will change over time as the organizations using Pulumi also change and evolve.

Reviewer 4: > it's important to note that these recommended practices can and will change over time as the organizations using Pulumi also change and evolve.

what does this mean? who does "organizations" refer to here? the hypothetical zephyr or you just saying that these won't always be the gold standard and these practices could be wrong if you read the post like 12 months from now?

Author: Looks like this could use some rewording. The intent behind this statement was to say that what might be "best" for any given customer at one point in time may not be "best" for that same customer 12 months later _if_ their circumstances have changed (where, like Zephyr, perhaps they've grown, perhaps they are using Pulumi for multiple initiatives, etc.). If you have suggestions on how to better phrase this, then please let me know!

Author: I've reworded this, see the version in e569521b and let me know what you think

[IaC Recommended Practices: Code Organization and Stacks](/blog/iac-recommended-practices-code-organization-and-stacks/)

[IaC Recommended Practices: Developer Stacks and Git Branches](/blog/iac-recommended-practices-developer-stacks-git-branches/)

**IaC Recommended Practices: Structuring Pulumi Projects** (this post)

IaC Recommended Practices: Local Testing with Pulumi

IaC Recommended Practices: Tying Stacks Together with Stack References

IaC Recommended Practices: Evolving the Application

IaC Recommended Practices: Adding Pulumi Deployments

IaC Recommended Practices: Refactoring for Reuse

## Catching up with Zephyr

As a result of launching its online presence, Zephyr saw amazing success. The market for arcane artifacts and novel curiosities is booming, and Zephyr is becoming a leader in this market. Funded by this success, Zephyr has grown rapidly. Along with this growth, Zephyr's software engineering and IT teams also grew and expanded. Fortunately, Zephyr's [use of per-developer stacks with short-lived ephemeral Git branches](/blog/iac-recommended-practices-developer-stacks-git-branches/) positioned the development teams well for this growth.

> 

Reviewer 4: I like how the 2nd post drops this as a note to the first post. we can do the same pointing to the 2nd post's code

{{% notes %}}
For a snapshot of the code as it was at the end of the first post in the series, see the [`original`](https://github.com/pulumi/zephyr-app/tree/original) tag of the [`pulumi/zepyhr-app`](https://github.com/pulumi/zephyr-app) repository on GitHub.
{{% /notes %}}

Author: Good idea, I'll update the post. We should just make this part of every post's structure moving forward.

Author: Addressed in e569521b

* **Resource lifecycle:** Do resources have a shared lifecycle, meaning they are always created, updated, or deleted together? (Note this is _not_ the same as a dependency on another resource!) If so, then those resources should be placed in the same project.
* **Change rate of resources:** Some resources may change very infrequently. Consider the example of an AWS VPC and associated objects. Once it is created, it may not change for extended periods of time, if ever. On the other hand, some resources may change frequently (in other words, these resources have a high change rate). In a single project, combining resources that have a high change rate with resources that have a low change rate may be suboptimal.

Now that you have an idea of some of the factors that affect how users structure their Pulumi projects, what are the advantages and disadvantages of each approach?

> For single projects, the key advantage is _simplicity._ There's no denying that a single project is simpler and less complex than working with multiple projects. This advantage, though, is also the key disadvantage. A single project does not afford the flexibility to let different teams manage different aspects of the infrastructure, for example, nor does it offer the ability to split resources that don't have a shared lifecycle. You can't run `pulumi destroy` on only _some_ of the resources in a project.

Reviewer 4: do we want to make the parallel here to monoliths vs microservices at the application level? its sort of the same concept, its all about the scaling action and the blast radius

Author: Yeah, that's a good idea! I'll update the post shortly.

Author: I've added this in e569521b

Reviewer 2: In practice, I think you're rarely running `pulumi destroy` on production infrastructure, and while you can't run `destroy` on just some of the infra, you can comment some of it out...

Author: @jkodroff OK, that's a fair point. Would you recommend just removing that sentence, or is there a better example to use to help illustrate the point?

Reviewer 2: I think we're really talking about blast radius here, so maybe just say something like "as we mentioned earlier, blast radius is a concern".

Finally, there is the consideration of what happens in the event of human error, and limiting the so-called "blast radius." With a single project, _all_ the resources in that project are within the blast radius---all of them could potentially be affected in the event of human error. Using separate Pulumi projects can limit the blast radius. It's not something that should drive the decision all by itself, but it _is_ a factor to be considered.

_In general, most customers end up using multiple Pulumi projects per use case._ As you've seen, though, there are a number of factors that influence that decision and will affect how many projects end up in use.

## Moving forward with Zephyr

Zephyr had a number of factors in play, not the least of which was the organizational change resulting from their growth and success. In the end, Zephyr decided to restructure their use of Pulumi for the online retail store into three separate projects:

* An infrastructure project: This project handled the AWS VPC and related infrastructure resources.
* A platform project: This project managed the Kubernetes cluster(s).

> * An application project: This project handled the actual online retail store itself.

Reviewer 4: earlier you mentioned only 2 teams. did you mean 3 teams?

Author: I mentioned three teams earlier, but I can see where that may not have been clear. Adjusted the wording earlier in the article in e569521b

{{% /notes %}}

In [the first Zephyr post](/blog/iac-recommended-practices-code-organization-and-stacks/), you read about some of the options for where to store Pulumi code relative to the applications it supports. At that time, Zephyr had decided to go with what's called a monorepo approach, meaning that the Pulumi code was stored in the same repository as the application code. In conjunction with the project reorganization, Zephyr also decided to change how their code was organized:

* Code for the infrastructure project moved to a separate "zephyr-infra" repository
* A new "zephyr-k8s" repository housed the code for the platform project
* The code for the online retail store application itself remained in the same "zephyr-app" repository

Note that relocating code (to a different GitHub repository or different filesystem location within a repository) generally has no impact on Pulumi---keeping in mind, of course, that filesystem paths and such may need to be adjusted in code to account for any such changes.

> These repositories are available for you to review on GitHub; use the branch/tag selector to find the `multi-project` tab in each repository to see the state of the code as of this blog post.

Reviewer 4: do you want to just hyperlink to it here?

Author: I will hyperlink to the repos here, but the repos don't exist yet. :-)

Author: I will add the hyperlinks after the repos are created but before the post goes live.

Reviewer: Agree a table or something with links to each would be nice here.

Reviewer: What do you think about maybe showing (or at least explaining) how these are (or will need to be) deployed now?

Author: I've added hyperlinks to the repositories, and added a note about checking the instructions in each repo to see how to deploy the online store now that it is in multiple projects.
